### PR TITLE
bug fix in ApplyAdadelta update rule

### PR DESCRIPTION
**Pull request by dpaiton**
Following the Zeiler ADADELTA paper, numerator of update equation
should be RMS(accum_update) \* grad

### CODE REVIEW

**Comment by dpaiton** [comment]
@vrv had requested that I also merge a fix to the test. I looked in [training_ops_test.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/training_ops_test.cc) and did not find any testing for adadelta. I also checked [adadelta_test.py](https://github.com/tensorflow/tensorflow/blob/16d395e5dea687ab3aece0a462e631de25c8d77d/tensorflow/python/training/adadelta_test.py), which has the correct rule in place. I do not know where else I should look for the bug.

**Comment by vrv** [comment]
Correct me if I'm wrong, but you're saying that adadelta_test has the right equation, but that this test was passing before your change, and still passes after your change?

**Comment by dpaiton** [comment]
Correct, the file [adadelta_test.py](https://github.com/tensorflow/tensorflow/blob/16d395e5dea687ab3aece0a462e631de25c8d77d/tensorflow/python/training/adadelta_test.py) passes before and after the fix.

**Comment by vrv** [comment]
doesn't that mean that the existing test is not really sufficient?  Can we modify the test so it would have failed before your fix?

**Comment by dpaiton** [comment]
Probably - I suspect that decreasing the tolerance in the asserts (from 1e-3 to a smaller number) would cause the test to fail, although I don't know if there is a testing convention in place for tolerance. Is there a reason 1e-3 was chosen in the first place, or is that a default value? These tests use numerical approximations, right? So at some point a low enough tolerance will always cause a failure.

**Comment by vrv** [comment]
There's no really good default, since every test has a different number of operations being performed.  Ideally we would test based off of expected number of ULP differences for every test, but that's somewhat impractical.

If you can find a setting that passes with your change and fails before it, that would be awesome.

**Comment by dpaiton** [comment]
I'll look into it and get back to you.

**Comment by dpaiton** [comment]
Ok, I added additional test variable options (learning rates, gradients, number of ADADELTA steps) in order to find a use case where the previously broken code fails and the newly patch code passes.

File: tensorflow/python/training/adadelta_test.py [hunk 1]
```diff
@@ -20,104 +20,90 @@
...
+                rho = 0.95
+                epsilon = 1e-8
+                adadelta_opt = tf.train.AdadeltaOptimizer(lr, rho, epsilon)
+                adadelta_update = adadelta_opt.apply_gradients(zip(
+                    [grads, grads], [var0, var1]))
+
+                tf.initialize_all_variables().run()
+
+                # Assign slots
+                slot = [None]*2
```
Comment (vrv) [inline tensorflow/python/training/adadelta_test.py]: space between ] and *  and 2

File: tensorflow/python/training/adadelta_test.py [hunk 2]
```diff
@@ -20,104 +20,90 @@
...
+                tot_update = 0
+                for run_no in range(num_updates):
+                    # Run adadelta update for comparison
+                    adadelta_update.run()
+
+                    # Perform initial update without previous accum values
+                    accum = accum * rho + (grad**2) * (1 - rho)
+                    update[run_no] = (np.sqrt(accum_update + epsilon) *
+                        (1. / np.sqrt(accum + epsilon)) * grad)
+                    accum_update = accum_update * rho + (update[run_no]**2) * (1.0 - rho)
```
Comment (vrv) [inline tensorflow/python/training/adadelta_test.py]: our linters go nuts when lines go over 80 characters, can you try to wrap these to 80 cols as necessary?)

File: tensorflow/python/training/adadelta_test.py [hunk 3]
```diff
@@ -20,104 +20,90 @@
...
-            var0.eval(),
-            rtol=1e-3)
-
-        self.assertAllCloseAccordingToType(
-            np.array([3.0 - update1 - update2, 4.0 - update1 - update2]),
-            var1.eval(),
-            rtol=1e-3)
-
+        for grad in [0.2, 0.1, 0.01]:
+            for lr in [1.0, 0.5, 0.1]:
```
Comment (vrv) [inline tensorflow/python/training/adadelta_test.py]: for better or worse, we use 2 space indents

File: tensorflow/python/training/adadelta_test.py [hunk 4]
```diff
@@ -20,104 +20,90 @@
...
+                self.assertEquals(slot_update[1].get_shape(), var1.get_shape())
+                self.assertFalse(slot_update[1] in tf.trainable_variables())
+
+                # Fetch params to validate initial values
+                self.assertAllClose(var0_init, var0.eval())
+                self.assertAllClose(var1_init, var1.eval())
+
+                update = [None]*num_updates
+                tot_update = 0
+                for run_no in range(num_updates):
```
Comment (vrv) [inline tensorflow/python/training/adadelta_test.py]: "step" is a more common parlance in TF code

**Comment by vrv** [comment]
Thanks for this!  mostly just lint / style issues, otherwise this looks great.

**Comment by dpaiton** [comment]
Ok, I fixed the linting/style errors. Should I squash the commits or can you do that in the merge?

**Comment by vrv** [comment]
We can squash now when we merge :)

@tensorflow-jenkins test this please

**Comment by martinwicke** [comment]
Thanks!

**Comment by dpaiton** [comment]
Thanks for your help @vrv
